database/schema/upgrade_scripts/122-vulnerable_package_tables.sql

```diff
@@ -0,0 +1,150 @@
+  id BIGSERIAL,
+  name TEXT NOT NULL UNIQUE, CHECK (NOT empty(name)),
+  PRIMARY KEY (id)
+) TABLESPACE pg_default;
+
+GRANT SELECT, INSERT, UPDATE, DELETE ON package_name TO ve_db_user_evaluator;
+GRANT SELECT, INSERT, UPDATE, DELETE ON package_name TO ve_db_user_vmaas_sync;
+GRANT SELECT, INSERT, UPDATE, DELETE ON package_name TO ve_db_user_advisor_listener;
+GRANT INSERT, UPDATE ON package_name TO ve_db_user_taskomatic;
+GRANT SELECT ON package_name TO ve_db_user_manager;
+
+-- cpe table
+CREATE TABLE cpe (
+  id BIGSERIAL,
+  name TEXT NOT NULL UNIQUE, CHECK (NOT empty(name)),
+  PRIMARY KEY (id)
+) TABLESPACE pg_default;
+
+GRANT SELECT, INSERT, UPDATE, DELETE ON cpe TO ve_db_user_evaluator;
+GRANT SELECT, INSERT, UPDATE, DELETE ON cpe TO ve_db_user_vmaas_sync;
+GRANT SELECT, INSERT, UPDATE, DELETE ON cpe TO ve_db_user_advisor_listener;
+GRANT INSERT, UPDATE ON cpe TO ve_db_user_taskomatic;
+GRANT SELECT ON cpe TO ve_db_user_manager;
+
+-- vulnerable package table
+CREATE TABLE vulnerable_package (
+  id BIGSERIAL,
+  package_name_id BIGINT NOT NULL,
+  cpe_id BIGINT NOT NULL,
+  UNIQUE(package_name_id, cpe_id),
+  PRIMARY KEY (id),
+  CONSTRAINT package_name_id
+    FOREIGN KEY (package_name_id)
+    REFERENCES package_name (id),
+  CONSTRAINT cpe_id
+    FOREIGN KEY (cpe_id)
+    REFERENCES cpe (id)
+) TABLESPACE pg_default;
+
+GRANT SELECT, INSERT, UPDATE, DELETE ON vulnerable_package TO ve_db_user_evaluator;
+GRANT SELECT, INSERT, UPDATE, DELETE ON vulnerable_package TO ve_db_user_vmaas_sync;
+GRANT SELECT, INSERT, UPDATE, DELETE ON vulnerable_package TO ve_db_user_advisor_listener;
+GRANT INSERT, UPDATE ON vulnerable_package TO ve_db_user_taskomatic;
+GRANT SELECT ON vulnerable_package TO ve_db_user_manager;
+
+-- vulnerable package cve table
+CREATE TABLE vulnerable_package_cve (
+  id BIGSERIAL,
+  vulnerable_package_id BIGINT NOT NULL,
+  cve_id BIGINT NOT NULL,
+  PRIMARY KEY (id),
+  CONSTRAINT vulnerable_package_id
+    FOREIGN KEY (vulnerable_package_id)
+    REFERENCES vulnerable_package (id),
+  CONSTRAINT cve_id
+    FOREIGN KEY (cve_id)
+    REFERENCES cve_metadata (id)
+) TABLESPACE pg_default;
+
+GRANT SELECT, INSERT, UPDATE, DELETE ON vulnerable_package_cve TO ve_db_user_evaluator;
+GRANT SELECT, INSERT, UPDATE, DELETE ON vulnerable_package_cve TO ve_db_user_vmaas_sync;
+GRANT SELECT, INSERT, UPDATE, DELETE ON vulnerable_package_cve TO ve_db_user_advisor_listener;
+GRANT INSERT, UPDATE ON vulnerable_package_cve TO ve_db_user_taskomatic;
+GRANT SELECT ON vulnerable_package_cve TO ve_db_user_manager;
+
+-- system vulnerable package table
+CREATE TABLE IF NOT EXISTS system_vulnerable_package (
+  id BIGSERIAL,
+  rh_account_id BIGINT NOT NULL,
+  system_id BIGINT NOT NULL,
+  vulnerable_package_id BIGINT NOT NULL,
+  first_reported TIMESTAMP WITH TIME ZONE NOT NULL,
+  status_id INT DEFAULT 0,
+  status_text TEXT,
+  state VULNERABILITY_STATE,
+  UNIQUE (system_id, vulnerable_package_id),
+  PRIMARY KEY (id),
```

Comment:
Let's avoid adding primary key here in parent table since it complains that the partitioning expression have to part of it. It's enough to set primary key only for each partition.

Mentioned [here](https://www.postgresql.org/docs/current/ddl-partitioning.html) as a limitation that PostgreSQL can't guarantee that id will be unique across all partitions:

> To create a unique or primary key constraint on a partitioned table, the partition keys must not include any expressions or function calls and the constraint's columns must include all of the partition key columns. This limitation exists because the individual indexes making up the constraint can only directly enforce uniqueness within their own partitions; therefore, the partition structure itself must guarantee that there are not duplicates in different partitions.

Reply:
Thanks, I was just looking for it!